### DISCUSSION TITLE
How to Turnkey Zeptoforth applications

### DISCUSSION THREAD

**Discussion by ctrilsbach** [tabemann/zeptoforth#14] [General]
Hi Travis,

How do you make a application run on power up ?

Thanks

Chris Trilsbach

**Comment by ctrilsbach** [comment]
Hi Travis,

Please ignore this discussion as I found the information I wanted in the Wiki page. Still getting used to Github.

Chris

**Comment by tabemann** [comment]
As I said in the other thread, any questions are good ones.

**Comment by tabemann** [comment]
@ctrilsbach, one thing I should note is that that information in the wiki you read is outdated, and I have since updated it. The recommended way to implement a turnkey now is to define a word named `turnkey`, which will be executed *after* `init` is executed, even if `init` is defined most recently after `turnkey` was implemented. This works in more recent versions of zeptoforth, and had not been implemented at the time I wrote that originally.

**Comment by ctrilsbach** [comment]
I tried this with version 0.54 but it just locks up and I cannot use Ctlr-c to get out.

Below is the part of my code that runs everything. It used to be a word called main but I replaced it with turnkey and initialised the variables that used to be initialised in main in init.

I tried compiling to ram and init works ok but turnkey produces a hard fault. Presumably turnkey is not supposed to be used this way.

: init init
0 poscntr !
0 posstart !
0 posend !
2 numimgs !
3 modecntr C! 
;

: turnkey
setup-display
clear-screen
modechange
0 ldcnt 
begin
	modebutton
	25 ms
	key?
until
offall
clear-screen
;

**Reply by tabemann** [reply]
Ctrl-C is a "best effort" attempt to reboot; if the system is sufficiently corrupted, even it won't work. There must be a problem somewhere in your `turnkey` that is severely corrupting the system state. I did just confirm locally on a Pico that `turnkey` itself works for me (by blinking an LED ten times before returning, which starts the REPL).

**Comment by ctrilsbach** [comment]
After a few attempts I have proved that it is nothing to do with turnkey or init. If I compile my code to ram it works fine but crashes with a hard fault if I compile it to flash and run it. 
I have realised that I don't understand how compile to flash works on Zeptoforth. I presumed that when code is compiled to flash it is just copied to ram on start up and then becomes exactly the same as if it was compiled to ram in the first case. This is certainly my experience of how Mecrisp works and you just have a init that calls your code. No variables need to be initialised.
Please can you explain why init has to call init and variables need to be initialised on Zeptoforth.
Thanks
Chris

**Reply by tabemann** [reply]
Chris,

zeptoforth differs from Mecrisp-Stellaris in that code compiled to flash runs from flash even on the RP2040, except for the kernel, which is copied to RAM on boot on the RP2040 (but runs from flash in STM32 MCU's).

One key thing about zeptoforth is that when code is compiled to flash, and `variable`s, `value`s, or `buffer:`s are defined, you must reboot zeptoforth before accessing them, because they are given fixed, compiled-in locations in RAM upon being compiled, but those locations are not available yet for use until you reboot. Not rebooting before accessing them will likely result in a crash or undefined behavior.

`init` must call `init` first thing because critical system initialization is carried out by the previously-defined instances of `init`; the system will not be properly initialized if `init` does not call `init` as the first thing it does, which will result in undefined behavior at best.

`variable`s are not automatically initialized in zeptoforth, unlike in Mecrisp-Stellaris; it is typical of Forths to not automatically initialize `variable`s, and Mecrisp-Stellaris is atypical in this regard (even though overall it is a rather traditional Forth). Note that `value`s, however, are automatically initialized in zeptoforth when `init` is executed.

Travis

**Comment by ctrilsbach** [comment]
Thanks Travis, that is a lot clearer.
Chris

**Reply by ctrilsbach** [reply]
Hi Travis,

I still cannot get my application to turnkey and as is just hangs it is difficult to debug. To be honest I gave up in frustration after reinstalling Zeptoforth about twenty times but would still like to get it to work as I put a lot of time in to it. I have initialised all my variables in init and can get small examples to turnkey fine but not my code for some reason. Is there anything else that could be causing the hang ?

**Reply by tabemann** [reply]
@ctrilsbach Could you paste your `init` and `turnkey` code? One important note is that you *must* start your `init` routine with a call to the previous definition of `init` or otherwise it will not work. Thanks.

**Reply by ctrilsbach** [reply]
Hi Travis,

This is my init and turnkey code.

: init init 
0 buttonstore ! 
0 poscntr !
0 posstart !
0 posend !
2 numimgs !
3 modecntr C! 
0 bytecntr !
0 charcntr !
;

: turnkey
setup-display
clear-screen
modechange
0 ldcnt 
begin
	modebutton
	25 ms
	key?
until
offall
clear-screen
;

**Reply by tabemann** [reply]
Chris,

Unless there are alignment issues, which there should not be if you are using `variable` and `cvariable`, your `init` code looks good to me. My guess is that something is failing inside `turnkey`, so the next thing I would try would be putting prints on each line of `turnkey` such as `." A"`, `." B"`, `." C"`, etc. to pick out what line it is failing on. (It is safe to print to and read from the serial console inside `turnkey`.) I would also try manually executing each word in `turnkey` by itself (aside from the `begin` and `until` obviously) to also help pick out the point of failure.

Travis

**Reply by ctrilsbach** [reply]
Travis,

I found that the line 

init-display 17 0 >i2c-stop

Is causing the problem. init-display is just an array of 17 chars to initialise the display.

I moved all the i2c initialisation to the init word and now it no longer crashes but the display is corrupted. 

I an using version 0.62.0 at the moment.

Chris

: init init 
0 buttonstore ! 
0 poscntr !
0 posstart !
0 posend !
2 numimgs !
3 modecntr C! 
0 bytecntr !
0 charcntr !

0 master-i2c
0 enable-i2c
0 4 i2c-pin
0 5 i2c-pin
4 pull-up-pin
5 pull-up-pin

0 7-bit-i2c-addr
$3C 0 i2c-target-addr! 
;

**Reply by tabemann** [reply]
One little note - `>i2c-stop` pushes the number of bytes actually transmitted so you will probably want to drop it.

**Reply by tabemann** [reply]
Note that when you first initialize an SSD1306-based display, the display will look corrupted, until you write actual image data to it. Of course, there are other possible causes of corrupted-looking screens, such as an incorrect SETCOMPINS setting (something that I ran into myself). I would highly suggest looking at the SSD1306 driver code in `extra/common/ssd1306.fs`.

**Reply by ctrilsbach** [reply]
Hi Travis,

I did drop it in the next line of code.

As to the display it all works fine when compiled to ram just not if compiled to flash ?

**Reply by tabemann** [reply]
@ctrilsbach How are you initializing `init-display`?

**Reply by tabemann** [reply]
@ctrilsbach If you are `create`ing `init-display` and then later using `c!` to write to it it will not work when compiled to flash, because `init-display` will live in flash and thus not be able to be written to with `c!`. If you `create` `init-display` initialize it with `c,`, or if you want `init-display` to live in RAM, define it with `buffer:` instead so it is alloted as a buffer in RAM on bootup.

**Reply by ctrilsbach** [reply]
Hi Travis, init-display is a read only array.

create init-display 
$00 c,  \ control byte for command stream
$A6 c,  \ set normal display
$D5 c, $80 c, \ set display clock divider
$8D c, $14 c, \ set charge pump
$81 c, $7F c, \ set contrast
$DA	c, $12 c, \ set com pin map
$AF c, \ set display on
$21 c, $00 c, $7F c, \ set max column 127
$22 c, $00 c, $03 c, \ set max pages 4

**Reply by tabemann** [reply]
@ctrilsbach That definitely looks correct to me. The big thing I wonder about is what is the difference between the code when executed manually while in RAM versus executed on bootup from flash. Would you mind putting your code as a whole up somewhere (whether GitHub or a pastebin or like) so I can get a closer look at it as a whole?

**Reply by ctrilsbach** [reply]
Yes sure. 
[Tslicerz.zip](https://github.com/tabemann/zeptoforth/files/11408555/Tslicerz.zip)

**Reply by tabemann** [reply]
@ctrilsbach From a cursory lookover of your code,  you are using `create` and `allot` to create buffers that are modified at runtime, such as `i2c-buf`. These should all be created with \<size> `buffer:` \<buffer-name> instead, which will create buffers of that size in RAM, because when compiling to flash `create` and `allot` will allot space in flash instead, which cannot be used for non-constant arrays. Also, `modecntr` ought to be a `cvariable` or `variable` (technically a `variable` will use more space, but it may in some cases avoid alignment issues, so I would recommend a `variable`  here).

Also, I notice a typo here:

```
$000000


00 , $3C3C3030 , $03030F0F , $00000000 ,
```

which should be:

```
$00000000 , $3C3C3030 , $03030F0F , $00000000 ,
```

to avoid leaving garbage on the stack.

Travis

**Reply by ctrilsbach** [reply]
Thanks Travis. 

That makes sense. I will change the code and try again.

Chris

**Reply by ctrilsbach** [reply]
Hi Travis,

Turnkey all working now.

Thanks again for your help.

Chris

**Comment by znmeb** [comment]
@ctrilsbach Are you on a platform that supports `gdb`? IIRC some of the `zeptoforth` implementations support `gdb`, and there is now an easy hardware setup for the Raspberry Pi Pico with the [Raspberry Pi Debug Probe](https://www.raspberrypi.com/documentation/microcontrollers/debug-probe.htmll). I buy them in pairs now - a Pico H and a debug probe. :-)

**Reply by ctrilsbach** [reply]
@znmeb Yes I am connecting to the Pico via a Raspberry Pi so gdb should be available.

**Reply by znmeb** [reply]
> @znmeb Yes I am connecting to the Pico via a Raspberry Pi so gdb should be available.

Sounds good. In case you haven't seen it, the instructions for setting it up are in here. <https://datasheets.raspberrypi.com/pico/getting-started-with-pico.pdf>

As I noted above I buy my Picos in pairs now. The Pico H has male headers soldered on and has a three-pin Serial Wire Debug (SWD) connector at the end of the board opposite the USB connector. The debug probe has a USB connector at one end and two three-pin connectors at the other end. One three-pin connector carries SWD data and the other carries serial port data.

So you can plug the Pico H into a breadboard, connect the debug probe to the Pico and access both the SWD data and serial port data from the Pico on the host through the debug probe's USB port.

**Reply by tabemann** [reply]
One note - gdb is not aware of zeptoforth's Forth symbols. I used gdb to help bring up zeptoforth from just a kernel, but it has not been of much use since then since I cannot easily find or call Forth words that are not part of the kerenel iself.

**Reply by ctrilsbach** [reply]
@znmeb Thanks for that. I will try Travis's suggestion first of printing before each line in turnkey first to see if I can find the problem that way. If not luck I will have a look at using GDB.

**Comment by mosconi** [comment]
Hi Travis.

I`m new to forth, and exploring it.  I would like to know how to compile the "blinky task example" with the init/turnkey, so the led blinks on when the rp2040 starts, while the console is available to use.

If I try to type the following, while in `compile-to-flash`, the RP stops and when I turn off and on the device, the led blinks, but the RP doesn't responds to the console(usb):
```
: turnkey 
  0 :noname begin green toggle-led 500 ms again ; 256 128 512 spawn blink-task !
  blink-task @ run
;
```
if I try to do ` 0 :noname begin green toggle-led 500 ms again ; 256 128 512 spawn blink-task !` in `init`, it doesn`t compile and on reboot, the led blinks, but the console(usb) is not working.

What am I doing wrong?

**Comment by tabemann** [comment]
On Sat, Aug 10, 2024 at 4:48 PM Rodrigo Mosconi ***@***.***>
wrote:

> Hi Travis.
>
> I`m new to forth, and exploring it. I would like to know how to compile
> the "blinky task example" with the init/turnkey, so the led blinks on when
> the rp2040 starts, while the console is available to use.
>
> If I try to type the following, while in compile-to-flash, the RP stops
> and when I turn off and on the device, the led blinks, but the RP doesn't
> responds to the console(usb):
>
> : turnkey
>   0 :noname begin green toggle-led 500 ms again ; 256 128 512 spawn blink-task !
>   blink-task @ run
> ;
>
> if I try to do 0 :noname begin green toggle-led 500 ms again ; 256 128
> 512 spawn blink-task ! in init, it doesn`t compile and on reboot, the led
> blinks, but the console(usb) is not working.
>
> What am I doing wrong?
>
The issue with your code is that you are using :NONAME ... ; inside another
word, in this case TURNKEY or INIT, when :NONAME ... ; cannot be nested
inside another word and will break horribly if you try. Rather what you
should be using is [: ... ;] , which is specifically designed to be nested
inside another word. Remember, :NONAME ... ; outside a word, [: ... ;]
inside a word.

I modified your code accordingly as:

compile-to-flash
task import
led import
variable blink-task
: turnkey
  0 [: begin green toggle-led 500 ms again ;] 256 128 512 spawn blink-task !
  blink-task @ run
;
reboot

As soon as I entered REBOOT the LED on my Pico started blinking, and when I
connected again with picocom to /dev/ttyACM0 (I am using an rp2040_big
full_usb build) I got:

Welcome to zeptoforth
Built for rp2040_big, version 1.7.4, on Wed Aug 7 08:11:39 PM CDT 2024
zeptoforth comes with ABSOLUTELY NO WARRANTY: for details type `license'
 ok

Hope this helps!

Travis

**Reply by mosconi** [reply]
Thanks !

That` s worked.  Could be possible to add that example to the wiki?
